core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -108,8 +108,78 @@ class TransactionStateManager(brokerId: Int,
     }
   }
 
-  def enablePidExpiration() {
-    // TODO: add producer id expiration logic
+
+
+  def enableTransactionalIdExpiration() {
+    scheduler.schedule("transactionalId-expiration", () => {
+      val now = time.milliseconds()
+      val(transactionalIdAndTransitionByPartition, recordsPerPartition) = inLock(stateLock) {
+        val transactionalIdByPartition: Map[Int, mutable.Iterable[TransactionalIdCoordinatorEpochAndMetadata]] =
+          transactionMetadataCache.flatMap { case (partition, entry) =>
+          entry.metadataPerTransactionalId.filter { case (_, txnMetadata) => txnMetadata.state match {
+            case Empty | CompleteCommit | CompleteAbort => true
+            case _ => false
+          }}.filter { case (_, txnMetadata) =>
+            txnMetadata.txnLastUpdateTimestamp <= now - config.transactionalIdExpirationMs
```

Comment:
since we are not grabbing a lock on the object, we should make `@volatile txnLastUpdateTimestamp` to make sure we get the latest value.